api/inc/register_gateway.h

```diff
@@ -17,105 +17,146 @@
-                                              __UVISOR_REGISTER_GATEWAY_METADATA2, \
-                                              __UVISOR_REGISTER_GATEWAY_METADATA1, \
-                                              /* No macro for 0 args */          )(src_box, ##__VA_ARGS__)
+/** `BX LR` encoding
+ * @internal
+ */
+#define BXLR                           ((uint16_t) 0x4770)
 
-/* Rregister-level gateway - Read */
-/* FIXME: Currently only a hardcoded 32bit constant can be used for the addr
- *        field. */
-#define uvisor_read(src_box, ...) \
+/** Register Gateway - Read operation
+ *
+ * This macro provides a variadic API to perform 32-bit read operations on
+ * restricted registers. Such accesses are assembled into a read-only flash
+ * structure that is read and validated by uVisor before performing the
+ * operation.
+ *
+ * The following APIs can be used:
+ * void uvisor_read(src_box, address);
+ * void uvisor_read(src_box, address, operation, mask);
+ *
+ * @warning These APIs currently only support link-time known values. This means
+ * that an access of this type is not supported:
+ * ```C
+ * uint32_t some_value = 1;
+ * uvisor_write(&SYSCFG->CTRL, some_value); // Yes
+ * uint32_t some_value = rand();
+ * uvisor_write(&SYSCFG->CTRL, some_value); // No
+ * ```
+ *
+ * @param box_name[in]  The name of the source box as decalred in
+ *                      `UVISOR_BOX_CONFIG`.
+ * @param addr[in]      The address for the data access.
+ * @param operation[in] The operation to perform at the address for the read. It
+ *                      is chosen among the `UVISOR_RGW_OP_*` macros.
+ * @param mask[in]      The mask to apply for the read operation.
+ * @returns The value read from address using the operation and mask provided
+ * (or their respective defaults if they have not been provided).
+ */
+#define uvisor_read(box_name, addr, ...) \
     ({ \
-        uint32_t res = UVISOR_SVC(UVISOR_SVC_ID_REGISTER_GATEWAY, \
-                                  __UVISOR_REGISTER_GATEWAY_METADATA(src_box, ##__VA_ARGS__)); \
-        res; \
+        /* Instanstiate the gateway. This gets resolved at link-time. */ \
+        static const TRegisterGateway register_gateway = { \
+            .svc_opcode = UVISOR_SVC_OPCODE(UVISOR_SVC_ID_REGISTER_GATEWAY), \
+            .branch     = BRANCH_OPCODE(OFFSETOF(TRegisterGateway, branch), OFFSETOF(TRegisterGateway, bxlr)), \
+            .magic      = UVISOR_REGISTER_GATEWAY_MAGIC, \
+            .box_ptr    = (uint32_t) & box_name ## _cfg_ptr, \
+            .address    = (uint32_t) addr, \
+            .value      = 0, \
+            .mask       = SELECT_MASK(__VA_ARGS__), \
+            .operation  = SELECT_READ_OP(__VA_ARGS__), \
+            .bxlr       = BXLR  \
+        }; \
+        \
+        /* Call the actual gateway. */ \
+        uint32_t result = ((uint32_t (*)(void)) ((uint32_t) &register_gateway | 1))(); \
```

Comment:
Yes. At the moment I've tried the following:

* Just cast to a fn pointer. Doesn't work (Thumb bit not set).
* Save `gtw_ptr | 1` in a `static const` and use that as fn ptr. Doesn't work (`gtw_ptr` not known until link time, the `| 1` is done at runtime.

Please feel free to suggest other approaches!